src/mail.rs

```diff
@@ -7,105 +7,115 @@ use std::path::Path;
-            subject: String::new(),
-            html: String::new(),
-            text: String::new(),
-            from_name: String::new(),
-            reply_to: String::new(),
-            date: String::new(),
+            from: "",
+            subject: "",
+            html: "",
+            text: "",
+            from_name: "",
+            reply_to: "",
+            date: "",
             attachments: HashMap::new(),
             content: HashMap::new(),
             headers: HashMap::new(),
-            x_smtpapi: String::new(),
+            x_smtpapi: "",
         }
     }
 
     /// Adds a CC recipient to the Mail struct.
-    pub fn add_cc<T: Into<String>>(&mut self, cc_addr: T) {
-        self.cc.push(cc_addr.into())
-    }
+    add_field!(add_cc << cc: &'a str);
 
     /// Adds a to recipient to the Mail struct.
-    pub fn add_to<T: Into<String>>(&mut self, to_addr: T) {
-        self.to.push(to_addr.into())
-    }
+    add_field!(add_to << to: Destination<'a>);
 
     /// Set the from address for the Mail struct. This can be changed, but there
     /// is only one from address per message.
-    pub fn add_from<T: Into<String>>(&mut self, from_addr: T) {
-        self.from = from_addr.into()
-    }
+    add_field!(add_from = from: &'a str);
 
     /// Set the subject of the message.
-    pub fn add_subject<T: Into<String>>(&mut self, subject: T) {
-        self.subject = subject.into()
-    }
+    add_field!(add_subject = subject: &'a str);
 
     /// This function sets the HTML content for the message.
-    pub fn add_html<T: Into<String>>(&mut self, html: T) {
-        self.html = html.into()
-    }
-
-    /// Add a name for the "to" field in the message. The number of to names
-    /// must match the number of "to" addresses.
-    pub fn add_to_name<T: Into<String>>(&mut self, to_name: T) {
-        self.to_names.push(to_name.into());
-    }
+    add_field!(add_html = html: &'a str);
 
     /// Set the text content of the message.
-    pub fn add_text<T: Into<String>>(&mut self, text: T) {
-        self.text = text.into()
-    }
+    add_field!(add_text = text: &'a str);
 
     /// Add a BCC address to the message.
-    pub fn add_bcc<T: Into<String>>(&mut self, bcc_addr: T) {
-        self.bcc.push(bcc_addr.into())
-    }
+    add_field!(add_bcc << bcc: &'a str);
 
     /// Set the from name for the message.
-    pub fn add_from_name<T: Into<String>>(&mut self, from_name: T) {
-        self.from_name = from_name.into()
-    }
+    add_field!(add_from_name = from_name: &'a str);
 
     /// Set the reply to address for the message.
-    pub fn add_reply_to<T: Into<String>>(&mut self, reply_to: T) {
-        self.reply_to = reply_to.into()
-    }
+    add_field!(add_reply_to = reply_to: &'a str);
 
     /// Set the date for the message. This must be a valid RFC 822 timestamp.
-    pub fn add_date(&mut self, date: String) {
-        self.date = date
+    // TODO(richo) Should this be a chronos::Utc ?
```

Comment:
Did you want to investigate this now or open an issue to adjust this later?

Reply:
Let's kick that can down the road for now. I think it boils down to rethinking how serialization of this form works. I had a crack at it in [this branch](https://github.com/richo/sendgrid-rs/commit/9ed5d1b3040e5d6b2e8c46559a47fe9dc7c9eabf) but ultimately ran into some issues that I didn't have time to followup on. I can open an issue for  it.